core/services/ocr2/plugins/mercury/encoding/decoder.go

```diff
@@ -0,0 +1,111 @@
+package encoding
+
+import (
+	"encoding/binary"
+
+	ocrtypes "github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
+	mercuryv1 "github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/v1"
+	mercuryv2 "github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/v2"
+	mercuryv3 "github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/v3"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+
+	"github.com/pkg/errors"
+)
+
+const REPORT_V1 uint16 = 1
+const REPORT_V2 uint16 = 2
+const REPORT_V3 uint16 = 3
+
+func DecodeSchemaVersionFromFeedId(feedID [32]byte) (uint16, error) {
+	schemaVersion := binary.BigEndian.Uint16(feedID[:2])
+	if schemaVersion != REPORT_V1 && schemaVersion != REPORT_V2 && schemaVersion != REPORT_V3 {
+		return 0, errors.Errorf("invalid schema version: %d", schemaVersion)
+	}
+	return schemaVersion, nil
+}
+
+type ReportDecoder interface {
+	GetSchemaVersion() uint16
+
+	DecodeAsV1() (*mercuryv1.Report, error)
+	DecodeAsV2() (*mercuryv2.Report, error)
+	DecodeAsV3() (*mercuryv3.Report, error)
+}
```

Comment:
This API is a big confusing. Why not expose pure functions instead?

```
encoding.DecodeV1([]byte report) (*mercuryv1.Report, error)
// etc
```

Or you may even be able to use generics here to make it quite clean.

Reply:
I couldn't find a way to use generics here unfortunately. Maybe I missed something?
The idea behind having an object was to consolidate the logic. The external user would have to know schema type first and then decode based on its value. So `SchemaVersion()` and `Decode()` always go together. And storing schema version allows to avoid decoding is again when `Decode()` is called.

Reply:
We can't use generics directly in the relayer API with LOOP Plugins, so that is not really a problem. However, I agree that it is confusing to construct a single use decoder that wraps one particular report. I think Sam is suggesting an API where a decoder would be multi-use, and passed the report in the method args instead. I think this makes sense, since the report type is just bytes, and the method calls can be efficient, without extra set-up for each invocation.